File: gobot/history.go
package main

import "strings"

type History struct {
	data []string
	size int
	pos  int
}

func NewHistory(size int) *History {
	return &History{
		data: make([]string, size),
		size: size,
		pos:  0,
	}
}

func (fs *History) Add(element string) {
	fs.data[fs.pos] = element
	fs.pos = (fs.pos + 1) % fs.size
}

func (fs *History) String() string {
	result := make([]string, fs.size)
	for i := 0; i < fs.size; i++ {
		result[i] = fs.data[(fs.pos+i)%fs.size]
	}
	return strings.Join(result, "\n")
}

File: rustbot/src/history.rs
use std::collections::VecDeque;
use std::fmt;

#[derive(Clone, Debug)]
pub struct History {
    data: VecDeque<String>,
    size: usize,
}

impl History {
    pub fn new(size: usize) -> Self {
        History {
            data: VecDeque::with_capacity(size),
            size,
        }
    }

    pub fn add(&mut self, element: String) {
        if self.data.len() == self.size {
            self.data.pop_front();
        }
        self.data.push_back(element);
    }

    pub fn string(&self) -> String {
        let mut s = String::new();
        for string in &self.data {
            s.push_str(string);
            s.push('\n');
        }
        s.pop(); // Remove the last newline character
        s
    }
}

impl fmt::Display for History {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        let mut result = Vec::with_capacity(self.size);
        for string in &self.data {
            result.push(string.as_str());
        }
        write!(f, "{}", result.join("\n"))
    }
}
